### ISSUE TITLE
Crash on SSLError handshake timeout

### ISSUE THREAD

**Issue by ryansb** [chmouel/ftp-cloudfs#27]
I have a bucket with 20k+ objects, and when trying to connect using version 0.14 (with the 10k+ object patch) I get an SSLError that closes my connection to the ftpcloudfs daemon. 

Stacktrace: http://pastebin.com/g6Ln9FAk

**Comment by reidrac** [comment]
Are you using the daemon with Rackspace Cloud Files or an OpenStack implementation?

**Comment by ryansb** [comment]
I'm using Rackspace Cloud Files, in the US (so I authenticate against https://auth.api.rackspacecloud.com/v1.0 )

**Comment by reidrac** [comment]
I think it may be related to this python-cloudfiles bug:

https://github.com/rackspace/python-cloudfiles/issues/34

The best thing we can do is try to catch these ssl.SSLError EVERYWHERE and retry transparently when possible, although it's a workaround to the python-cloudfiles problem (that I think it's probably a problem with Python's SSL code).

If you can use Rackspace CF without SSL (not the authentication, only the communication with the storage once you got a token), that would work as a temporal workaround.

Thanks for your report.

**Comment by ryansb** [comment]
I see. I have an exponential backoff decorator, and here it is adapted to this purpose. Do you think I could send you a pull request with this decorator implemented to catch and silently retry SSL errors? 

```
def retry(maxtries=5, delay=.25, backoff=2, message=None, success_message=None):
    """Retry decorator with exponential backoff
    Retries a function or method until it returns True.

    maxtries: must be at least 0
    delay: sets the initial delay in seconds. Must be greater than 0
    backoff: sets the factor by which the delay should lengthen after each failure
        (backoff must be greater than 1, or else it isn't really a backoff)
    message: a message to print after each attempt, optional
    """
    if backoff <= 1:
        raise ValueError("backoff must be greater than 1")
    maxtries = int(maxtries)
    if maxtries < 0:
        raise ValueError("tries must be 0 or greater")
    if delay <= 0:
        raise ValueError("delay must be greater than 0")

    def deco_retry(func):
        def f_retry(*args, **kwargs):
            mtries, mdelay = maxtries, delay  # make mutable
            ret_val = func(*args, **kwargs)  # first attempt
            if ret_val:  # If it's successful the first time, just gtfo
                return True
            while mtries > 0:
                if ret_val:  # Done on success
                    if success_message:
                        print success_message
                    return True
                if message:
                    print message
                mtries -= 1      # consume an attempt
                time.sleep(mdelay)  # wait...
                mdelay *= backoff  # make future wait longer
                ret_val = func(*args, **kwargs)  # Try again
            return False  # Ran out of tries :-(
        return f_retry  # true decorator -> decorated function
    return deco_retry  # @retry(arg[, ...]) -> true decorator
```

**Comment by reidrac** [comment]
The problem is in python-cloudfiles, I think the best solution would be sub-classing Connection. I don't want to introduce code everywhere that we may need to remove eventually because the problem is fixed upstream :)

We already have a ProxyConnection class that adds the X-Forwarded-For header to all the requests. I think the best place to handle that is there, in make_request: https://github.com/chmouel/ftp-cloudfs/blob/master/ftpcloudfs/fs.py#L38 (if you check your traceback, that's our entry point to python-cloudfs code)

Basically we can catch ssl.SSLErrors (and anything else not catch by python-cloudfiles Connection) and retry accordingly.

What do you think?

EDIT: I ate a couple of words.

**Comment by ryansb** [comment]
That seems reasonable, I'll put together a pull request and do some tests on it later today. 

Do you think I should implement backoffs, or nix those in favor of just waiting a quarter of a second and trying again?

**Comment by reidrac** [comment]
Take a look to the original make_request: https://github.com/rackspace/python-cloudfiles/blob/master/cloudfiles/connection.py#L160

I think you don't need to wait at all. The ssl.SSLError is fatal :) meaning that it need to open a new connection (it calls to http_connect method). Basically retry n times and then convert the ssl.SSLError to something IOSError (tip: look at translate_cloudfiles_error ; do it there!).

To be honest, it would be better if that ssl.SSLError was added to that line: https://github.com/rackspace/python-cloudfiles/blob/master/cloudfiles/connection.py#L151

Adding this as a workaround it's all right, but we should fix it in python-cloudfiles.

Give it a go, I don't want to spoil your fun!

**Comment by ryansb** [comment]
I'll make a pull request for this project, and I'll do one for
python-cloudfiles, though it seems like there's a lot in the pull request
queue for cloudfiles, so it could be a while before it goes mainline.

Edit: My mail client inserted a bunch of cruft.

**Comment by reidrac** [comment]
Thanks that would be great!

**Comment by reidrac** [comment]
I've committed a change that should solve this issue,  SHA: 5f015de5e68a5268bcdc2daccdbe48e768468079 and SHA: 
6b0d7e2413ad3c1682f204b1e8331fed291fbbaf

Please reference this bug report when submitting the patch to python-cloudfiles so I can close it.

Thanks again for your report!

**Comment by reidrac** [comment]
I've released version 0.15 that includes a fix for this.
